mlx/io/gguf.cpp

```diff
@@ -89,27 +209,178 @@ std::unordered_map<std::string, array> load_gguf(
     const auto& [data, dtype] = extract_tensor_data(&tensor);
     array loaded_array = array(data, shape, dtype);
     std::string name = std::string(tensor.name, tensor.namelen);
-    result.insert({name, loaded_array});
+    array_map.insert({name, loaded_array});
   }
+  return array_map;
+}
+
+std::pair<
+    std::unordered_map<std::string, array>,
+    std::unordered_map<std::string, MetaData>>
+load_gguf(const std::string& file, StreamOrDevice s) {
+  gguf_ctx* ctx = gguf_open(file.c_str());
+  if (!ctx) {
+    throw std::runtime_error("[load_gguf] gguf_init failed");
+  }
+  auto metadata = load_metadata(ctx);
+  auto arrays = load_arrays(ctx);
   gguf_close(ctx);
-  return result;
+  return {arrays, metadata};
+}
+
+void append_kv_array(
+    gguf_ctx* ctx,
+    const std::string& key,
+    array& val,
+    uint32_t gguf_type) {
+  if (val.ndim() == 1) {
+    size_t gguf_size = val.nbytes() + gguf_array_header;
+    std::vector<char> val_vec(gguf_size);
+    gguf_value* gguf_val = reinterpret_cast<gguf_value*>(val_vec.data());
+    gguf_val->array.type = gguf_type;
+    gguf_val->array.len = val.size();
+    memcpy(val_vec.data() + gguf_array_header, val.data<char>(), val.nbytes());
+    gguf_append_kv(
+        ctx,
+        key.c_str(),
+        key.length(),
+        GGUF_VALUE_TYPE_ARRAY,
+        reinterpret_cast<void*>(val_vec.data()),
+        gguf_size);
+  } else {
```

Comment:
is this ever the case? Reading the code, it seems that we call `gguf_append_kv` directly if we have a string or an array of strings.

Reply:
This case is to handle scalars (0 dim) arrays differently than 1 d arrays.